applets/wncklet/window-list.c

```diff
@@ -139,6 +146,166 @@ static void applet_change_background(MatePanelApplet* applet, MatePanelAppletBac
+
+	/* Generate window screenshot for preview */
+	screenshot = gdk_pixbuf_get_from_window (window, 0, 0, width / scale, height / scale);
+	g_object_unref (window);
+
+	if (screenshot == NULL)
+		return NULL;
+
+	/* Determine whether the contents of the screenshot are empty */
+	pixels = gdk_pixbuf_get_pixels (screenshot);
+	if (!g_strcmp0 ((const char *)pixels, ""))
+	{
+		g_object_unref (screenshot);
+		return NULL;
+	}
+
+	/* Scale to configured size while maintaining aspect ratio */
+	if (width > height)
+	{
+		ratio = (double) height / (double) width;
+		width = MIN(width, tasklist->thumbnail_size);
+		height = width * ratio;
+	}
+	else
+	{
+		ratio = (double) width / (double) height;
+		height = MIN(height, tasklist->thumbnail_size);
+		width = height * ratio;
+	}
+
+	thumbnail = gdk_pixbuf_scale_simple (screenshot, width, height, GDK_INTERP_BILINEAR);
+	g_object_unref (screenshot);
+
+	return thumbnail;
+}
+
+#define PREVIEW_PADDING 5
+static void preview_window_reposition (TasklistData *tasklist, GdkPixbuf *thumbnail)
+{
+	GdkMonitor *monitor;
+	GdkRectangle monitor_geom;
+	int x_pos, y_pos;
+	int width, height;
+
+	width = gdk_pixbuf_get_width (thumbnail);
+	height = gdk_pixbuf_get_height (thumbnail);
+
+	/* Resize window to fit thumbnail */
+	gtk_window_resize (GTK_WINDOW (tasklist->preview), width, height);
+
+	/* Set position at pointer, then re-adjust from there to just outside of the pointer */
+	gtk_window_set_position (GTK_WINDOW (tasklist->preview), GTK_WIN_POS_MOUSE);
+	gtk_window_get_position (GTK_WINDOW (tasklist->preview), &x_pos, &y_pos);
+
+	/* Get geometry of monitor where tasklist is located to calculate correct position of preview */
+	monitor = gdk_display_get_monitor_at_point (gdk_display_get_default (), x_pos, y_pos);
+	gdk_monitor_get_geometry (monitor, &monitor_geom);
+
+	/* Add padding to clear the panel */
+	switch (mate_panel_applet_get_orient (MATE_PANEL_APPLET (tasklist->applet)))
+	{
+		case MATE_PANEL_APPLET_ORIENT_LEFT:
+			x_pos = monitor_geom.width + monitor_geom.x - (width + tasklist->size) - PREVIEW_PADDING;
+			break;
+		case MATE_PANEL_APPLET_ORIENT_RIGHT:
+			x_pos = tasklist->size + PREVIEW_PADDING;
+			break;
+		case MATE_PANEL_APPLET_ORIENT_UP:
+			y_pos = monitor_geom.height + monitor_geom.y - (height + tasklist->size) - PREVIEW_PADDING;
+			break;
+		case MATE_PANEL_APPLET_ORIENT_DOWN:
+		default:
+			y_pos = tasklist->size + PREVIEW_PADDING;
+			break;
+	}
+
+	gtk_window_move (GTK_WINDOW (tasklist->preview), x_pos, y_pos);
+}
+
+static gboolean preview_window_draw (GtkWidget *widget, cairo_t *cr, GdkPixbuf *thumbnail)
+{
+	GtkStyleContext *context;
+
+	context = gtk_widget_get_style_context (widget);
+	gtk_render_icon (context, cr, thumbnail, 0, 0);
+
+	return FALSE;
+}
+
+static gboolean applet_enter_notify_event (WnckTasklist *tl, GList *wnck_windows, TasklistData *tasklist)
+{
+	GdkPixbuf *thumbnail;
+	WnckWindow *wnck_window = NULL;
+	int n_windows;
+
+	if (GTK_IS_WIDGET (tasklist->preview))
```

Comment:
Can preview be something that is not widget?

Reply:
Without it I kept getting a bunch of these errors:
```
(wnck-applet:17578): Gtk-CRITICAL **: 09:04:29.058: gtk_widget_destroy: assertion 'GTK_IS_WIDGET (widget)' failed
```

Reply:
You are destroying preview twice. First in `applet_leave_notify_event` then in `applet_enter_notify_event`...